### DISCUSSION TITLE
Is it possible to pass HttpClientFactory to KubernetesClinet (IKubernetes implementation) for creating HttpClient?

### DISCUSSION THREAD

**Discussion by sankar1204** [kubernetes-client/csharp#1364] [Q&A]
I want to pass my own httpClientFactory to the IKubernetes instance, so is it possible to pass HttpClientFactory while creating instance of IKubernetes ?

As I explored it, sometime in 2019 - 2021 it was allowed to take HttpClient in constructor of Kubernetes, but currently its only taking DelegatingHandler in args. Is there any way that I can pass IhttpclientFactroy now.

If someone knows why HttpClient arg is removed from constructor args please explain, or redirect me to that page.

**Comment by tg123** [comment]
need to install watcher handler to httpclient

**Reply by sankar1204** [reply]
can you please direct me to an example where httpClient is passed from outside to KubernetesClient

**Reply by tg123** [reply]
not support
may i know your scenario?

why do you want your own httpclient

**Reply by sankar1204** [reply]
I need to 
- add some default requestHeaders to the request sent from KubernetesClient
- modify retries on certain failure response codes

**Reply by tg123** [reply]
inherit kubernetes class and override `SendRequest` 

https://github.com/kubernetes-client/csharp/blob/e64c372e34785f75bc66da21938ab84a164a2996/src/KubernetesClient.Basic/AbstractKubernetes.cs#L104

**Reply by sankar1204** [reply]
let me try and get back. thanks

**Comment by roblappcoupa** [comment]
What happened with this? Why was the `Kubernetes` constructor that took an `HttpClient` removed? I need to create new instances of `Kubernetes` because the underlying file-based configuration changes a lot due to a dynamically generated authentication tokens. Therefore, I need to new up a `Kubernetes` objects often to ensure updated configuration is captured, but I don't want to create new `HttpClient` instances every time. I was hoping to control the `HttpClient` that was passed to the `Kubernetes` constructor but it was removed.

**Reply by tg123** [reply]
see reason why httpclient is hidden 
https://github.com/kubernetes-client/csharp/issues/165

**Reply by roblappcoupa** [reply]
Thanks for the link. Do you have any suggestions on the best way to manage the `HttpClient` instances so they come from `HttpClientFactory` to avoid socket exhaustion issue from creating too many instances?

**Reply by tg123** [reply]
any reason to create many instances?

**Reply by roblappcoupa** [reply]
It has to do with how the authentication is managed between my application and my Kubernetes cluster. My `.kube` config file looks like this:
```
users:
- name: arn:aws:eks:us-east-1:myawsaccount:cluster/mycluster
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - --region
      - us-east-1
      - eks
      - get-token
      - --cluster-name
      - mycluster
      - --output
      - json
      command: aws
```
With this, it's my understanding that a token for the Kubernetes API server is dynamically retrieved and it's only good for 15 minutes. Therefore, I need to create new instances of `Kubernetes` at least every 15 minutes so that they pick up the updated configuration. So my code basically creates a new `Kubernetes` instance for each workflow that interacts with the Kubernetes API. I could introduce caching so that it only creates one every 15 minutes which would result in fewer instances, but I wanted to try managing the `HttpClient` instances through `HttpClientFactory` instead.

**Reply by tg123** [reply]
sdk will refresh token when expires, you do not have to create a new instance

**Reply by roblappcoupa** [reply]
After reading your comment and going through the source code I see that now. This changes things for me, so you can disregard the question as I don't need to control the HttpClient instances as I won't need to create so many instances of `Kubernetes` once I refactor my code. Thanks for the quick responses and guidance!
